### ISSUE TITLE
Get default setting value from settingtypes.txt

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#6711]
Currently it's impossible to get the default value of a setting in the Lua API.

So I propose to add a function (or extend the existing ones) to grab the default value of a setting (as specified in `settingtypes.txt`).
For invalid queries, or if the setting does not have a default in `settingtypes.txt`, return `nil`.

**Comment by dsohler** [comment]
I second that. It is an absolute must-have and would avoid mods implementing something like this over and over again:

```Lua
local get_value = function (setting, default_value)
    local s = minetest.settings:get(setting)
    return (s == nil and default_value or s)
end
```

Or even better: Load all `settingtypes.txt` files and return the value set there if it was not set by the user and thus not being in `minetest.conf`.

**Comment by rubenwardy** [comment]
Your snippet has redundant code

```lua
minetest.settings:get(setting) or default_setting
```

**Comment by dsohler** [comment]
Yeah, whatever. Anything relevant to say about the issue?

**Comment by rubenwardy** [comment]
That is fucking relevant, it deceptive to use unoptimal code to make the problem look bigger than it is

**Comment by dsohler** [comment]
> That is fucking relevant, it's deceptive to use inoptimal code to make the problem look bigger than it is

Then fucking implement it properly to the API instead of being an annoying prick against me every time I write something. Gosh … I can’t tell how much I hate you personally.

**Comment by rubenwardy** [comment]
Sure, this would be a nice addition, but I also consider it low/mediun priority as the alternative without it is a few characters more in mods. It's better to use Dev time on fixing the many bugs in the API, such as still present unintialised variables.

Feeling is mutual

**Comment by paramat** [comment]
dsohler please be nicer and less abrasive, if some people are losing patience with you it's due to your own behaviour, the only moderation option we have is to completely ban you from the Github minetest repos, we'd rather not.

Concerning this request, most settings are gettable because they are stored in .conf, this issue is for things like default noise params or mapgen specific params (like mgv7 floatland y) which are not stored in settings. So yes it's low priority.
See my code in `default/mapgen.lua`, i have to include the default values as a fallback:
```
-- Get setting or default
local mgv7_spflags = minetest.get_mapgen_setting("mgv7_spflags") or
	"mountains, ridges, nofloatlands, caverns"
local captures_float = string.match(mgv7_spflags, "floatlands")
local captures_nofloat = string.match(mgv7_spflags, "nofloatlands")

-- Get setting or default
-- Make global for mods to use to register floatland biomes
default.mgv7_floatland_level =
	minetest.get_mapgen_setting("mgv7_floatland_level") or 1280
default.mgv7_shadow_limit =
	minetest.get_mapgen_setting("mgv7_shadow_limit") or 1024
```

**Comment by Wuzzy2** [comment]
Hardcoding core setting default values into Lua is a terrible idea, modders should not be forced to do this.

**Comment by paramat** [comment]
I'm not suggesting that approach currently used is acceptable, i'm just showing what we have to do currently as a workaround.
I somewhat support this request.

**Comment by paramat** [comment]
Related #9284

**Comment by rubenwardy** [comment]
I think it's a good idea for defaults from settingtypes.txt to be used easily from `minetest.settings`. But how should compatibility be kept?

Mods may depend on `get()` returning nil if the user hasn't set any settings, and it's generally useful for mods to know if the user has set any settings. It would also be good to have some consistency between the engine and mods

**Comment by GoodClover** [comment]
If I'm correct engine settings (non-mod) *do* give their default if not set?
I can't think of a use for mods wanting to test if an option is set, and doubt any do… [XKCD 1172](https://xkcd.com/1172/)
I'd change `get` to return the default if not set, and nil if there's no default/not in settingtypes. Add an `is_set` for the sake of it.

For non-`minetest.settings` there should be the ability to give a settingtypes-format file in addition to the conf itself, i.e. `Settings("mysettings.conf", "mysettingtypes.txt")`.

**Comment by doxygen-spammer** [comment]
I want to bump this, because I consider this important.

My arguments:

* Having to supply the default value in code instead of settingtypes.txt violates the DRY rule. (don’t-repeat-yourself)
* Documentation is actually wrong [^1], it does not reflect the current facts.
  - lua_api.md mentions that `minetest.settings` makes either custom or existing settings accessible, which is vague, but indicates that there are two sources of settings being merged.
    https://github.com/minetest/minetest/blob/0ade097e99704fda81792b4e0853d1005997bc68/doc/lua_api.md#L255-L256
  - The documentation for Settings:get() does not say much. But for Settings:has() it says that minetest.settings:get() can return _some_ default values.
    https://github.com/minetest/minetest/blob/0ade097e99704fda81792b4e0853d1005997bc68/doc/lua_api.md#L8088-L8093
  - settingtypes.txt uses the term “default” without explaining it.
    The term “default” is well known in the field of software settings.
    It is the value being assumed by the software, in case no other value has been configured.
    What we have in settingtypes.txt is actually not a “default”, but a “placeholder”.
    Consider this screenshot:
    ![Screenshot_20230702_185033](https://github.com/minetest/minetest/assets/76133492/f8b2501c-b86a-419f-8ba6-20ee5b70d184)
    The text box already contains a string, but it is _not_ the configured value.
    Only after clicking “Set”, the value is used, but the red arrow indicates that it is _not_ the default anymore.
    Compare it with the excellent documentation for leading GUI toolkits: https://doc.qt.io/qt-6/qlineedit.html#placeholderText-prop, https://html.spec.whatwg.org/multipage/input.html#attr-input-placeholder
    Now it is clear that the term “default” is wrong, it is a “placeholder” instead.

And finally a suggestion about backwards compatibility:

There are already minetest.settings:get() and `minetest.setting_get_pos()`.
**I suggest to add minetest.setting_get(), which returns exactly what settingtypes.txt indicates.**
(`true` or `false` for bool settings, a string for string settings, a number for int settings, ..., and the default value from settingtypes.txt in case no other value is configured.)
It would be a very important function, so I think it is worth expanding the API here.

[^1]: My prior comment about “vague” vs. “misleading” vs. “wrong” applies here.

**Comment by Montandalar** [comment]
The only mods that I can see being broken by changing this is badly-made mods that already have their default value in their Lua code out of sync with what they have in settingtypes.txt. If fixing the API to deliver what people expect (which is what I think this issue really is about), then breaking a few badly-made mods along the way is acceptable collateral, and people who have to go fix them will wonder why it ever worked in the first place. Am I being too harsh?

**Comment by doxygen-spammer** [comment]
> badly-made mods that already have their default value in their Lua code out
> of sync with what they have in settingtypes.txt.
> [...] breaking a few badly-made mods along the way is acceptable
> [...] Am I being too harsh?

It’s too harsh, but I still agree.
These mods just made mistakes while being forced to violate DRY, that doesn’t 
make them badly-made.

But supporting such DRY mistakes will just provoke more mistakes of the same 
type (until all modders have learned to fix them), so nothing is lost by 
fixing the get() method.

Another thing pointed out was “checking for default value”, with functions 
like these:

```lua
local function is_default(key)
    return minetest.settings:get("my_mod_" .. key) == nil
end
```

```lua
local function is_default(key)
    return not minetest.settings:has_key("my_mod_" .. key)
end
```

Here I have to note that both methods are broken, we can’t break them anymore.
Because configuration items may explicitly store the default value, there is 
no way to figure out whether this is intentional or not.

**Comment by Wuzzy2** [comment]
Yes I agree. The main problem with the lack of such functionality is redundancy. It is bad mods are forced to redundantly write the default value in their code because what if the REAL default value in settingtypes.txt changes?

This is the main reason why I posted this issue to begin with.

**Comment by Xeno333** [comment]
Aren't engine defaults currently all hardcoded?? This is a terrible system. The engine should load all settings from the appropriate `settingtypes.txt` files for engine and mods.

If breaking compatability with old mods is a concern, we can simply pass two returns from `Settings(...):get(key)` such that it returns `value, default_value` thus `value = nil` would still be returned if it is not set and the default would also be showen. Again, this is only needed if compatability, MUST be kept, but again most mods shouldn't have out-dated settings in their code and new/active ones should be able to update.

**Comment by appgurueu** [comment]
> Yes I agree. The main problem with the lack of such functionality is redundancy. It is bad mods are forced to redundantly write the default value in their code because what if the REAL default value in settingtypes.txt changes?

I agree the engine should help, but there is something you can do as-is, which is to have your settingtypes declared in a Lua file from which you derive all the rest (`settingtypes.txt`, documentation on them, default values).
